### DISCUSSION TITLE
I noticed the MauiMaps app Android CustomMapHandler method MapPins() is being called twice for each Add

### DISCUSSION THREAD

**Discussion by davefxy** [VladislavAntonyuk/MauiSamples#313] [General]
I added a couple of debug calls in the Android CustomMapHandler method MapPins() in the MauMaps sample. I noticed that when I Add a location to the Map, the MapPins method being called twice in a row with the same map.Pins. This causes a double marker entry for each existing pin.
I would like to help determine why this is happening.
To debug this I downloaded the maui Net7.0 repository. I read the dotnet maui Development guide to learn how to use a local `bin\dotnet` scenario. I was able to compile the maui and start VS 2022 MauiMaps project. I set a breakpoint in MapPins() method with the hope of determining where the 2 calls to MapPin() originated.
However all I accomplished was to test MauiMaps app with the latest maui net7.0 branch. I was not able to view where the two calls for MapPin() originated in Maui.
What additional steps in compiling .Net maui repository do I need to perform to be able to step through .Net maui code using the MauiMaps app for debugging?
Thank you for your patience.

**Comment by davefxy** [comment]
I was able to look at this double call to MapPins() by compiling .net Maui as described in dotnet maui Development guide and using  Maui.Controls.Sample PinItemsSourceGallery page.
MapPins() is also called twice in the Android Map Handler. But the markers are cleared before adding the Pins. So we don't see a duplication of markers. I don't know why this doesn't work like this in the MauiMaps sample.
Here is an explanation of why MapPins() is being called twice unnecessarily.
Following the code through src\Controls\Maps\Map.cs, the OnItemsSourceCollectionChanged() method is called. In this method the 			e.Apply(insert: (item, _, __) => CreatePin(item)) is executed. This causes a NotifyCollectionChangedAction Apply() which calls PinsOnCollectionChanged(), and then Handler?.UpdateValue(nameof(IMap.Pins)) => until MapPins().
Execution returns to OnItemsSourceCollectionChanged() to the line : Handler?.UpdateValue(nameof(IMap.Pins) => until MapPins().
And this is why MapPins() is being called twice for simply adding a Pin. I don't think the code was the intended to follow this path. Perhaps there should be a check here to prevent Handler?.UpdateValue(nameof(IMap.Pins) from being called twice, once in PinsOnCollectionChanged and then in OnItemsSourceCollectionChanged()?

**Reply by VladislavAntonyuk** [reply]
Thank you for your investigation @davefxy. It's better to ask this question in dotnet/maui repo. 

As for Markers. Don't use Markers.Clear(). You need to go through all Markers collection and for each marker call Remove method.

The best option is moving my code with pins to .net maui repo.

**Comment by davefxy** [comment]
Your latest changes have fixed all of the outstanding issues with the handling of the map pins. Excellent work. I am glad you took the time to determine the appropriate changes to get it all working. Thank you for doing that.
